examples/java8/src/test/java/org/apache/beam/examples/SnippetsTest.java

```diff
@@ -0,0 +1,111 @@
+  @Test
+  public void testCoGroupByKeyTuple() throws IOException {
+    // [START CoGroupByKeyTupleInputs]
+    final List<KV<String, String>> emailList = Arrays.asList(
+        KV.of("amy", "amy@example.com"),
+        KV.of("carl", "carl@example.com"),
+        KV.of("julia", "julia@example.com"),
+        KV.of("carl", "carl@email.com"));
+
+    final List<KV<String, String>> phoneList = Arrays.asList(
+        KV.of("amy", "111-222-3333"),
+        KV.of("james", "222-333-4444"),
+        KV.of("amy", "333-444-5555"),
+        KV.of("carl", "444-555-6666"));
+    // [END CoGroupByKeyTupleInputs]
+
+    // [START CoGroupByKeyTupleOutputs]
+    final TupleTag<String> emailsTag = new TupleTag();
+    final TupleTag<String> phonesTag = new TupleTag();
+
+    final List<KV<String, CoGbkResult>> expectedResults = Arrays.asList(
+        KV.of("amy", CoGbkResult
+          .of(emailsTag, Arrays.asList("amy@example.com"))
+          .and(phonesTag, Arrays.asList("111-222-3333", "333-444-5555"))),
+        KV.of("carl", CoGbkResult
+          .of(emailsTag, Arrays.asList("carl@email.com", "carl@example.com"))
+          .and(phonesTag, Arrays.asList("444-555-6666"))),
+        KV.of("james", CoGbkResult
+          .of(emailsTag, Arrays.asList())
+          .and(phonesTag, Arrays.asList("222-333-4444"))),
+        KV.of("julia", CoGbkResult
+          .of(emailsTag, Arrays.asList("julia@example.com"))
+          .and(phonesTag, Arrays.asList())));
+    // [END CoGroupByKeyTupleOutputs]
+
+    PCollection<String> actualFormattedResults =
+        Snippets.coGroupByKeyTuple(p, emailsTag, phonesTag, emailList, phoneList);
+
+    // [START CoGroupByKeyTupleFormattedOutputs]
+    final List<String> formattedResults = Arrays.asList(
+        "amy; ['amy@example.com']; ['111-222-3333', '333-444-5555']",
+        "carl; ['carl@email.com', 'carl@example.com']; ['444-555-6666']",
+        "james; []; ['222-333-4444']",
+        "julia; ['julia@example.com']; []");
+    // [END CoGroupByKeyTupleFormattedOutputs]
+
+    // Make sure that both 'expectedResults' and 'actualFormattedResults' match with the
+    // 'formattedResults'. 'expectedResults' will have to be formatted before comparing
+    List<String> expectedFormattedResultsList = new ArrayList<String>(expectedResults.size());
+    for (KV<String, CoGbkResult> e : expectedResults) {
+      String name = e.getKey();
+      Iterable<String> emails = e.getValue().getAll(emailsTag);
+      Iterable<String> phones = e.getValue().getAll(phonesTag);
+      String formattedResult = Snippets.formatCoGbkResults(name, emails, phones);
+      expectedFormattedResultsList.add(formattedResult);
+    }
+    PCollection<String> expectedFormattedResultsPColl =
+        p.apply(Create.of(expectedFormattedResultsList));
+    PAssert.that(expectedFormattedResultsPColl).containsInAnyOrder(formattedResults);
+    PAssert.that(actualFormattedResults).containsInAnyOrder(formattedResults);
+
+    p.run().waitUntilFinish();
```

Comment:
<!--new_thread; commit:b7cd18485fb9d3c89bc5177ad8ab8742faa39c06; resolved:0-->
Since you're using `TestPipeline`, this should be performed automatically

Reply:
When I remove this, I get this error:
`SnippetsTest.testCoGroupByKeyTuple » PipelineRunMissing The pipeline has not been run.`

Am I missing something?

Reply:
I mean the `waitUntilFinish`. You should still run your pipeline.